examples/3d-tiles/tileset-3d-layer.js

```diff
@@ -40,48 +40,132 @@ export default class Tileset3DLayer extends CompositeLayer {
+
+    let zoomMap = Math.max(Math.min(zoom, max), min);
+    zoomMap = (zoomMap - min) / (max - min);
+    zoomMap = Math.max(Math.min(1.0 - zoomMap, 1), 0);
+
+    // setup frameState so that tileset-3d-traverser can do it's job
+    const frameState = {
+      camera: {
+        position: cameraPosition,
+        direction: cameraDirection,
+        up: cameraUp,
+      },
+      height: height,
+      frameNumber: tick,
+      distanceMagic: zoomMap * zoomMagic, // zoom doesn't seem to update accurately? like it stays at the same number after a scroll wheel tick
+      sseDenominator: 1.15, // Assumes fovy = 60 degrees
+      /*******************************************
+        From cesium:
+        frustum.fov = CesiumMath.toRadians(60.0);
+        frustum._fovy = (frustum.aspectRatio <= 1) ? frustum.fov : Math.atan(Math.tan(frustum.fov * 0.5) / frustum.aspectRatio) * 2.0;
+        frustum._sseDenominator = 2.0 * Math.tan(0.5 * frustum._fovy);
+       *******************************************/
+    };
+
+    tileset3d.update(frameState, DracoWorkerLoader);
+
+    // Add layer for any renderable tile
+    const selectedTiles = tileset3d._selectedTiles;
+    for (const tile of selectedTiles) {
+      this._updateLayer(tile, frameState);
+    }
+
+    this._selectLayers(frameState);
+    console.log(zoom + ' ' + this.state.layers.length);
   }
 
-  async _loadTileset(tilesetUrl) {
-    const {depthLimit} = this.props;
-    let tileset3d = null;
-    if (tilesetUrl) {
-      const tilesetJson = await load(tilesetUrl);
-      tileset3d = new Tileset3D(tilesetJson, tilesetUrl);
-      tileset3d.traverse(tileHeader => this._loadTile3D(tileHeader), depthLimit);
+  // Grab only those layers who were seleted this frame
+  _selectLayers(frameState) {
+    const {layerMap} = this.state;
+    const {frameNumber} = frameState;
+    const selectedLayers = [];
+    const layerMapValues = Object.values(layerMap);
+    for (const value of layerMapValues) {
+      if (value.selectedFrame === frameNumber) {
+        selectedLayers.push(value.layer);
+      }
     }
-    this.setState({tileset3d});
+    this.setState({
+      layers: selectedLayers,
+      lastUpdateStateTick: frameNumber
+    });
   }
 
-  async _loadTile3D(tileHeader) {
-    await tileHeader.loadContent(DracoWorkerLoader);
+  // Layer is created and added to the map if it doesn't exist in map
+  // If it exists, update its selected frame.
+  _updateLayer(tileHeader, frameState) {
+    // Check if already in map
+    let {layerMap} = this.state;
+    const {lastUpdateStateTick} = this.state;
+    const {frameNumber} = frameState;
+    if (tileHeader.contentUri in layerMap) {
+        const value  = layerMap[tileHeader.contentUri];
+        if (tileHeader._selectedFrame === frameNumber && value.selectedFrame === lastUpdateStateTick) { // Was rendered last frame and needs to render again
```

Comment:
Wonder if this logic could go into the traverser, so that we can keep the layer as clean as possible (and make it easier to use in non-deck.gl frameworks).

Reply:
Pushed this into tileset.update. After the traversal I take I look at what tiles need to be rendered (_selectedTiles) and check any of these does not have gpu resource for it and push it onto a public array of tiles so that the rendering app can setup resources for those only. The rendering app should pass in a function that can test if a tile needs its resource bound to gpu  given a map of tile's uri -> some object that it knows about. Right now I'm just sticking the layerMap and the test function on the frameState object.

Is there a better way to check if a layer needs to be regenerated for deck? Right now it checks if it wasn't rendered last time using frameNumber and lastFrameNumber. I didn't see anything obvious on the layer object looking in the debugger.